Mathlib/Data/Nat/Choose/Mchoose.lean

```diff
@@ -0,0 +1,116 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir & María-Inés de Frutos—Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María-Inés de Frutos—Fernández
+-/
+
+import Mathlib.Algebra.BigOperators.Group.Finset
+import Mathlib.Algebra.Ring.Regular
+import Mathlib.Tactic.Ring.RingNF
+
+/-! # Multiple choose
+
+* `Nat.mchoose m n` is the number of (unordered) partitions of a set
+with `m * n` elements into `m` parts of `n`-element subsets
+
+This integer is defined as a finite product.
+
+* `Nat.mchoose_eq` shows that it equals `(m * n).factorial / (m.factorial * n.factorial ^ m)`
+and `Nat.mchoose_mul_factorial_mul_pow_factorial` establishes the equality
+  `mchoose m n * m.factorial * n.factorial ^ m = (m * n).factorial`.
+
+* `Nat.mchoose_mul_add` and `Nat.choose_mul_right` are lemmas for `Nat.choose`
+that couldn't lie in `Mathlib.Data.Nat.Choose.Basic` because their proof uses the `ring` tactic.
+
+
+## Note :
+
+* It has nothing to do with `Nat.multichoose`.
+
+* It is a particular case of a more general counting problem, counting the number
+of unordered paritions of a set with given “type” (a multiset).
+
+-/
+
+namespace Nat
+
+theorem choose_mul_add (m) {n : ℕ} (hn : n ≠ 0) :
+    (m * n + n).choose n = (m + 1) * (m * n + n - 1).choose (n - 1) := by
+  rw [← mul_left_inj' (mul_ne_zero (factorial_ne_zero (m * n)) (factorial_ne_zero n))]
+  set p := n - 1
+  have hp : n = p + 1 := (succ_pred_eq_of_ne_zero hn).symm
+  simp only [hp, add_succ_sub_one]
+  calc
+    (m * (p + 1) + (p + 1)).choose (p + 1) * ((m * (p+1))! * (p+1)!)
+      = (m * (p + 1) + (p + 1)).choose (p + 1) * (m * (p+1))! * (p+1)! := by ring
+    _ = (m * (p+ 1) + (p + 1))! := by rw [add_choose_mul_factorial_mul_factorial]
+    _ = ((m * (p+ 1) + p) + 1)! := by ring_nf
+    _ = ((m * (p + 1) + p) + 1) * (m * (p + 1) + p)! := by rw [factorial_succ]
+    _ = (m * (p + 1) + p)! * ((p + 1) * (m + 1)) := by ring
+    _ = ((m * (p + 1) + p).choose p * (m * (p+1))! * (p)!) * ((p + 1) * (m + 1)) := by
+        rw [add_choose_mul_factorial_mul_factorial]
+    _ = (m * (p + 1) + p).choose p * (m * (p+1))! * (((p + 1) * (p)!) * (m + 1)) := by ring
+    _ = (m * (p + 1) + p).choose p * (m * (p+1))! * ((p + 1)! * (m + 1)) := by rw [factorial_succ]
+    _ = (m + 1) * (m * (p + 1) + p).choose p * ((m * (p + 1))! * (p + 1)!) := by ring
```

Comment:
I think this is more easily done using `Nat.succ_mul_choose_eq`.

Reply:
Indeed.